### PR TITLE
[ADD] DCOS-1002  Create a TeamCity integrated test for Windows version of DCOS CLI

### PR DESCRIPTION
**Pull request by YuraTolochkevich**

### CODE REVIEW

**Comment by ssk2** [comment]
Let's squash all of these commits please :)

**Comment by YuraTolochkevich** [comment]
@ssk2  Thanks, ready

**Comment by swartzrock** [comment]
Squashed, tests passed. Sunil can you please merge if it looks good? LGTM.

**Comment by mgummelt** [comment]
Out of curiosity, is anyone actually using Windows?

File: cli/tests/integrations/common.py [hunk 1]
```diff
@@ -30,7 +30,7 @@ def exec_command(cmd, env=None, stdin=None):
         stderr=subprocess.PIPE,
         env=env)
 
-    stdout, stderr = process.communicate()
+    stdout, stderr = [_.replace(b'\r', b'') for _ in process.communicate()]
```
Comment (mgummelt) [inline cli/tests/integrations/common.py]: underscore is for variables you don't use.  you use this one.  name it something else

File: bin/clean.ps1 [hunk 2]
```diff
@@ -0,0 +1,5 @@
+﻿$BaseDir = (Get-Location).Path
```
Comment (mgummelt) [inline bin/clean.ps1]: Let's put these windows build files in their own subdirectory

**Comment by mgummelt** [comment]
For the existing tests, our entry point was bin/start_tests.sh  What is the entry point for this new test suite?

**Comment by ssk2** [comment]
@mgummelt, a not-insignificant proportion of enterprise users use Windows, including some of Mesosphere's customers. It was an important requirement for the CLI.

File: cli/tests/integrations/test_analytics.py [hunk 3]
```diff
@@ -115,64 +173,6 @@ def test_config_reporting_false():
         assert requests.post.call_count == 0
 
 
-@_mock
-def test_production_setting_true():
```
Comment (mgummelt) [inline cli/tests/integrations/test_analytics.py]: If you've just moved these, please move them back so we know nothing has changed

**Comment by mgummelt** [comment]
Which customers?

File: cli/tests/integrations/test_config.py [hunk 4]
```diff
@@ -12,20 +12,24 @@
 
 @pytest.fixture
 def env():
-    return {
+    r = os.environ.copy()
```
Comment (mgummelt) [inline cli/tests/integrations/test_config.py]: why this change?

File: cli/tests/integrations/common.py (hunk 1 above)
Comment (mgummelt) [inline cli/tests/integrations/common.py]: Please add a comment here describing why you're doing this.

File: cli/tests/integrations/test_dcos.py [hunk 5]
```diff
@@ -71,9 +71,11 @@ def test_version():
 
 
 def test_missing_dcos_config():
-    env = {
+    env = os.environ.copy()
```
Comment (mgummelt) [inline cli/tests/integrations/test_dcos.py]: why this change?

File: cli/tox.win.ini [hunk 6]
```diff
@@ -0,0 +1,41 @@
+[tox]
```
Comment (mgummelt) [inline cli/tox.win.ini]: why do we need a separate tox for windows?

File: dcos/package.py [hunk 7]
```diff
@@ -943,6 +944,25 @@ def copy_to_cache(self, target_dir):
                 'Unable to fetch packages from [{}]'.format(self.url))
 
 
+def onerror(func, path, exc_info):
```
Comment (mgummelt) [inline dcos/package.py]: please follow the pydoc documentation format we use for all other functions.  Also, if this function is private to this module, prefix it with a `_`

Comment (mgummelt) [inline dcos/package.py]: why did you create this function?

File: dcos/package.py [hunk 8]
```diff
@@ -985,8 +1005,14 @@ def validate(self):
 
         # TODO(CD): implement these checks in pure Python?
         scripts_dir = os.path.join(self._base_path, 'scripts')
-        validate_script = os.path.join(scripts_dir, '1-validate-packages.sh')
-        result = subprocess.call(validate_script)
+        suffix = 'ps1' if util.is_windows_platform() else 'sh'
+        validate_script = os.path.join(scripts_dir,
+                                       '1-validate-packages.' + suffix)
+        cmd = [validate_script]
```
Comment (mgummelt) [inline dcos/package.py]: do an if..else here to be more functional

File: cli/tests/integrations/common.py (hunk 1 above)
Comment (ytalashkocv) [inline cli/tests/integrations/common.py]: This is needed to get rid of '\r' from Windows's lines endings.

File: cli/tox.win.ini (hunk 6 above)
Comment (ytalashkocv) [inline cli/tox.win.ini]: Because Windows requires additional library pypiwin32

File: cli/tests/integrations/test_config.py (hunk 4 above)
Comment (ytalashkocv) [inline cli/tests/integrations/test_config.py]: Because previous version of code does not worked for Windows

File: cli/tests/integrations/test_dcos.py (hunk 5 above)
Comment (ytalashkocv) [inline cli/tests/integrations/test_dcos.py]: same as above

File: bin/clean.ps1 (hunk 2 above)
Comment (ytalashkocv) [inline bin/clean.ps1]: done

File: cli/tests/integrations/test_analytics.py (hunk 3 above)
Comment (ytalashkocv) [inline cli/tests/integrations/test_analytics.py]: Because of tests order is very important for Windows. With old order OSError raises (reproducible only on py3.4).

File: dcos/package.py (hunk 8 above)
Comment (ytalashkocv) [inline dcos/package.py]: yes, done

File: dcos/package.py (hunk 7 above)
Comment (ytalashkocv) [inline dcos/package.py]: This function for adding write permission to files on Windows platform. Because exception raised when trying to delete read-only files on Windows.

**Comment by ytalashkocv** [comment]
@mgummelt Please, re review

File: cli/tests/integrations/test_analytics.py (hunk 3 above)
Comment (mgummelt) [inline cli/tests/integrations/test_analytics.py]: Why does an OSError raise?  The tests should behave independently, so if they're not, we need to fix that.

File: cli/tests/integrations/test_config.py (hunk 4 above)
Comment (mgummelt) [inline cli/tests/integrations/test_config.py]: Yes, I assume that's the case for all the changes in this PR.  My question is *why* did it not work on Windows?

File: cli/tests/integrations/common.py [hunk 9]
```diff
@@ -30,7 +30,8 @@ def exec_command(cmd, env=None, stdin=None):
         stderr=subprocess.PIPE,
         env=env)
 
-    stdout, stderr = process.communicate()
+    stdout, stderr = [std_stream.replace(b'\r', b'')
```
Comment (mgummelt) [inline cli/tests/integrations/common.py]: I mean put a comment in the code

**Comment by mgummelt** [comment]
I still need the answer to this question:

For the existing tests, our entry point was bin/start_tests.sh What is the entry point for this new test suite?

File: dcos/package.py [hunk 10]
```diff
@@ -943,6 +944,25 @@ def copy_to_cache(self, target_dir):
                 'Unable to fetch packages from [{}]'.format(self.url))
 
 
+def _onerror(func, path, exc_info):
```
Comment (mgummelt) [inline dcos/package.py]: This still doesn't follow the pydoc format used in other functions.  Please follow that format.

Comment (mgummelt) [inline dcos/package.py]: Also this needs to be named something more specific such as `_rmtree_on_error`

Comment (mgummelt) [inline dcos/package.py]: Why is the ~/.git directory created read-only?  If our user clones the repo on windows, shouldn't the user have write access?

File: dcos/package.py [hunk 11]
```diff
@@ -943,6 +944,25 @@ def copy_to_cache(self, target_dir):
...
+    If the error is for another reason it re-raises the error.
+
+    Usage : ``shutil.rmtree(path, onerror=onerror)``
+    """
+    import stat
+    if not os.access(path, os.W_OK):
+        os.chmod(path, stat.S_IRWXU | stat.S_IRWXG | stat.S_IRWXO)
+        func(path)
+    else:
+        raise
```
Comment (mgummelt) [inline dcos/package.py]: Are you sure this works?  I think calling `raise` without an argument only works inside an `except` block

File: cli/tests/integrations/test_analytics.py (hunk 3 above)
Comment (ytalashkocv) [inline cli/tests/integrations/test_analytics.py]: We doesn't know exact reason, only think that this connected with some state on Windows. Fails in winapi function call:

Comment (mgummelt) [inline cli/tests/integrations/test_analytics.py]: This is an error in `test_marathon`, so why did you move tests in `test_analytics`?

Comment (ytalashkocv) [inline cli/tests/integrations/test_analytics.py]: This error raises in every test after starts at `test_analytics` module: 
https://teamcity.mesosphere.io/viewLog.html?buildId=21743&buildTypeId=ClosedSource_DcosCli_WindowsTests_WindowsTest&tab=buildLog#_focus=23637

**Comment by ytalashkocv** [comment]
About this question https://github.com/mesosphere/dcos-cli/pull/218#issuecomment-111558163:
Added `win_bin\start_tests.ps1` script

File: cli/tests/integrations/test_config.py (hunk 4 above)
Comment (YuraTolochkevich) [inline cli/tests/integrations/test_config.py]: Windows requires full copy of the ENV, with next its update. If we put new  poor Env like: {
        constants.PATH_ENV: os.environ[constants.PATH_ENV],
        constants.DCOS_CONFIG_ENV: os.path.join("tests", "data", "dcos.toml"),
        cli_constants.DCOS_PRODUCTION_ENV: 'false'
    } it doesn't work for Windows, and we got next error: 
STDOUT:
STDERR: Traceback (most recent call last):
 File "C:\Users\admin\tmp\dcos-cli\cli\.tox\py27-integration\Scripts\dcos-script.py", line 5, in <module>
   from pkg_resources import load_entry_point
 File "C:\Users\admin\tmp\dcos-cli\cli\.tox\py27-integration\lib\site-packages\pkg_resources\__init__.py", line 37, in <module>
   import email.parser
 File "c:\python27\Lib\email\parser.py", line 12, in <module>
   from email.feedparser import FeedParser
 File "c:\python27\Lib\email\feedparser.py", line 27, in <module>
   from email import message
 File "c:\python27\Lib\email\message.py", line 16, in <module>
   import email.charset
 File "c:\python27\Lib\email\charset.py", line 13, in <module>
   import email.base64mime
 File "c:\python27\Lib\email\base64mime.py", line 40, in <module>
   from email.utils import fix_eols
 File "c:\python27\Lib\email\utils.py", line 27, in <module>
   import random
 File "c:\python27\Lib\random.py", line 885, in <module>
   _inst = Random()
 File "c:\python27\Lib\random.py", line 97, in __init__
   self.seed(x)
 File "c:\python27\Lib\random.py", line 113, in seed
   a = long(_hexlify(_urandom(2500)), 16)
WindowsError: [Error -2146893795] Provider DLL failed to initialize correctly

File: dcos/package.py (hunk 11 above)
Comment (ytalashkocv) [inline dcos/package.py]: `raise` works fine. But I changed `raise` to `raise exc_info[0], exc_info[1], exc_info[2]`. This looks more obvious.

**Comment by ytalashkocv** [comment]
@mgummelt All comments have been addressed. Please review.

File: dcos/package.py [hunk 12]
```diff
@@ -943,6 +946,25 @@ def copy_to_cache(self, target_dir):
                 'Unable to fetch packages from [{}]'.format(self.url))
 
 
+def _rmtree_on_error(func, path, exc_info):
```
Comment (mgummelt) [inline dcos/package.py]: This *still* doesn't follow our pydoc format.   Please look at the docstrings for the other functions.

File: dcos/package.py (hunk 11 above)
Comment (mgummelt) [inline dcos/package.py]: I just tried it in python2.7, and got this exception:

`TypeError: exceptions must be old-style classes or derived from BaseException, not NoneType`

Let's be explicit about what we're raising

File: cli/tests/integrations/test_analytics.py (hunk 3 above)
Comment (mgummelt) [inline cli/tests/integrations/test_analytics.py]: Why do you think moving the tests fixes it?  Why did you try that?

I'd really like to know why this is happening before we commit this.  If we don't understand this error, then in the likely case that it comes up again, we're not going to be able to fix it.

File: dcos/package.py (hunk 12 above)
Comment (ytalashkocv) [inline dcos/package.py]: @mgummelt Improve `_rmtree_on_error` pydoc

File: cli/tests/integrations/test_analytics.py (hunk 3 above)
Comment (ytalashkocv) [inline cli/tests/integrations/test_analytics.py]: When this failure happened (tests failed on py3.4 but works fine for py2.7), we noticed that tests have different order for py2.7 and py3.4(py3.4 runs tests in same order as they are ordered in python module). So we tried to change tests order (as they runned by py2.7). And this helped. Also we try to use py.test annotations like `@pytest.mark.first` but they doesn't work. As I mentioned here https://github.com/mesosphere/dcos-cli/pull/218#discussion_r32337010, we doesn't know exact reason of this failure, but if needed we can research this problem (we also tried, but without any success).

File: dcos/package.py (hunk 11 above)
Comment (ytalashkocv) [inline dcos/package.py]: You got such `TypeError: exceptions must be old-style classes or derived from BaseException, not NoneType` message, because, in current call stack, exception raising not happened (`raise` without arguments re-raises last raised exception). In our case `_rmtree_on_error` is an onerror callback, that called only after exception raised.

**Comment by ytalashkocv** [comment]
@mgummelt All comments addressed. Please review.

File: cli/tests/integrations/test_analytics.py (hunk 3 above)
Comment (mgummelt) [inline cli/tests/integrations/test_analytics.py]: Yes, please research the cause.  This is likely to come up again as we evolve the test suite.

File: cli/tests/integrations/test_config.py (hunk 4 above)
Comment (mgummelt) [inline cli/tests/integrations/test_config.py]: Hmm.   This might be OK, but can you first figure out which missing env var is causing this failure?

File: dcos/package.py (hunk 11 above)
Comment (mgummelt) [inline dcos/package.py]: Ah, got it.  Let's go ahead and raise exc_info anyway, so this function doesn't have to depend on global state (like the exception stack)

Comment (ytalashkocv) [inline dcos/package.py]: `raise` without arguments - is a common way for re-raising exceptions. One more reason, is changes in py3 for exception raising. If raise exception manually using exc_info, then such `During handling of the above exception, another exception occurred` message will be shown. Py3 identifying this exception like new, and not like re-raising old one.

Comment (mgummelt) [inline dcos/package.py]: OK

File: cli/tests/integrations/test_config.py (hunk 4 above)
Comment (ytalashkocv) [inline cli/tests/integrations/test_config.py]: It's `SYSTEMROOT` env variable. Value of this variable is `C:\Windows`.

File: cli/tests/integrations/test_analytics.py (hunk 3 above)
Comment (ytalashkocv) [inline cli/tests/integrations/test_analytics.py]: I do not found any additional information about this issue.

**Comment by ytalashkocv** [comment]
@mgummelt Comments addressed. Please review.

**Comment by mgummelt** [comment]
OK, one last thing: Who setup the teamcity configuration for this?

I'm very concerned that the failing windows test doesn't actually have anything to do with test ordering, and might just be a non-deterministic bug.  Regardless, it's a failure that we don't understand, and it's likely to come up again.

**Comment by ytalashkocv** [comment]
@mgummelt I have setup teamcity configuration and Yurii Tolochkevych create ami image (Windows server 2012).

**Comment by mgummelt** [comment]
How is DCOS installed on that machine?  From the TC config, it looks to be a long-running DCOS instance that you've preinstalled.  I think this might cause problems with frameworks re-registering in subsequent test runs because ZooKeeper state isn't cleared.

We run DCOS in a vagrant image from the dcos-image project for our other tests.  Have you looked into that?

**Comment by mgummelt** [comment]
Regardless, this LGTM.  You can merge.  I'll leave it up to you to manage that TC config.  Let me know if you need any help.

**Comment by YuraTolochkevich** [comment]
@mgummelt, Right, we use preinstalled  DCOS cluster for tests. Regarding DCOS vagrant  image,  we tried to deploy DCOS with vagrant, but our AWS  Windows machine is Virtual machine itself and it is a case of the nested visualization, and only allows deploying  32-bit  VMs with Virtual Box, VmWare or other provider. If there is an possibility to provide Windows machine installed on bare metal, we can try run tests with vagrant

**Comment by mgummelt** [comment]
@spahl ^^
